src/org/omegat/core/team2/impl/GITCredentialsProvider.java

```diff
@@ -90,23 +90,32 @@ protected void configure(OpenSshConfig.Host host, Session session) {
 
             @Override
             protected JSch createDefaultJSch(FS fs) throws JSchException {
-                Connector con = null;
-                try {
-                    if (SSHAgentConnector.isConnectorAvailable()) {
-                        USocketFactory usf = new JNAUSocketFactory();
-                        con = new SSHAgentConnector(usf);
-                    } else {
-                        ConnectorFactory cf = ConnectorFactory.getDefault();
-                        con = cf.createConnector();
+           		final JSch jsch = new JSch();
+                final File home = fs.userHome();
+                if (home != null) {
```

Comment:
Thanks for looking into this.

I understand that this is some setup boilerplate that is probably mostly equivalent to what we were doing before, but

- Is this documented anywhere?
- Do we really have to do all this? The library won't look for known hosts and supported private keys on its own?

Reply:
Sorry I should also say that this level of manual configuration still seems wrong to me.

I have my keys named after the service I created them for, like `~/.ssh/gitlab{,.pub}`. Enumerating supported key names just doesn't make sense to me. Shouldn't the library be transparently consulting `~/.ssh/config`?

Reply:
JGit/JSch looks `~/.ssh/config` and when configuration is there, it respect it.
I also have a habit to generate keys with name and configure `~/.ssh/config` with it.
The code here has no effect when user put key configuration.
It is just for default.

> Is this documented anywhere?

No. I understand things from JGit source code.

> Do we really have to do all this?

I want to override the behavior doing less than original, so I need to do it.
JSch don't have `removeIdentity` method, so I don't want to call super as previously did. 
And utility functions, such as knownHosts` are private so I need to write it.

The chunk of code doing similar things as [JGit method with same name](https://git.eclipse.org/r/plugins/gitiles/jgit/jgit/+/refs/tags/v5.11.1.202105131744-r/org.eclipse.jgit.ssh.jsch/src/org/eclipse/jgit/transport/JschConfigSessionFactory.java#399) because it override it.

```
	protected JSch createDefaultJSch(FS fs) throws JSchException {
		final JSch jsch = new JSch();
		JSch.setConfig("ssh-rsa", JSch.getConfig("signature.rsa")); //$NON-NLS-1$ //$NON-NLS-2$
		JSch.setConfig("ssh-dss", JSch.getConfig("signature.dss")); //$NON-NLS-1$ //$NON-NLS-2$
		configureJSch(jsch);
		knownHosts(jsch, fs);
		identities(jsch, fs);
		return jsch;
	}
	private static void knownHosts(JSch sch, FS fs) throws JSchException {
		final File home = fs.userHome();
		if (home == null)
			return;
		final File known_hosts = new File(new File(home, ".ssh"), "known_hosts"); //$NON-NLS-1$ //$NON-NLS-2$
		try (FileInputStream in = new FileInputStream(known_hosts)) {
			sch.setKnownHosts(in);
		} catch (FileNotFoundException none) {
			// Oh well. They don't have a known hosts in home.
		} catch (IOException err) {
			// Oh well. They don't have a known hosts in home.
		}
	}
	private static void identities(JSch sch, FS fs) {
		final File home = fs.userHome();
		if (home == null)
			return;
		final File sshdir = new File(home, ".ssh"); //$NON-NLS-1$
		if (sshdir.isDirectory()) {
			loadIdentity(sch, new File(sshdir, "identity")); //$NON-NLS-1$
			loadIdentity(sch, new File(sshdir, "id_rsa")); //$NON-NLS-1$
			loadIdentity(sch, new File(sshdir, "id_dsa")); //$NON-NLS-1$
		}
	}
	private static void loadIdentity(JSch sch, File priv) {
		if (priv.isFile()) {
			try {
				sch.addIdentity(priv.getAbsolutePath());
			} catch (JSchException e) {
				// Instead, pretend the key doesn't exist.
			}
		}
	}
```

Reply:
Sorry JSch have `remoteIdentity` method. I'd like to rewrite here.

Reply:
After rewriting `createDefaultJSch` method, OmegaT failed to clone. Now investigating...

Reply:
[JGit original `createDefaultJSch` method](https://git.eclipse.org/r/plugins/gitiles/jgit/jgit/+/refs/tags/v5.11.1.202105131744-r/org.eclipse.jgit.ssh.jsch/src/org/eclipse/jgit/transport/JschConfigSessionFactory.java#399) is like

```java
 	protected JSch createDefaultJSch(FS fs) throws JSchException {
	       final JSch jsch = new JSch();
               JSch.setConfig("ssh-rsa", JSch.getConfig("signature.rsa")); //$NON-NLS-1$ //$NON-NLS-2$
               JSch.setConfig("ssh-dss", JSch.getConfig("signature.dss")); //$NON-NLS-1$ //$NON-NLS-2$
```

Forked version of JSch already have "ssh-rsa" config, but it does not have "signature.rsa", "signature.dsa" so JSch.getConfig("signature.rsa") become null .
It override "ssh-rsa" configuration by null, then JSch class throws NPE at finally and failed to clone repository.

These two lines are merged at ["Ensure that JSch knows HostKeyAlgorithms ssh-rsa and ssh-dss"](https://git.eclipse.org/r/plugins/gitiles/jgit/jgit/+/09753591fae875cee71eb5674a0103844351e508)
which seems intend to work around original JSch problem.

The easiest way to avoid it is to implement whole functions by ourselves, which I did before.

Reply:
The issue has already reported to forked jsch project.

- https://github.com/mwiede/jsch/issues/85

It is also filed on JGit project.

- [Bug 576604 - JGit should not overwrite JSch properties, if they are already set.](https://bugs.eclipse.org/bugs/show_bug.cgi?id=576604)
- [Bug 576605 - Replace original JSch with an updated Fork](https://bugs.eclipse.org/bugs/show_bug.cgi?id=576605)